fabric-tag-api-v1/src/main/java/net/fabricmc/fabric/impl/tag/TagAliasLoader.java

```diff
@@ -0,0 +1,176 @@
+import java.io.Reader;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import com.google.gson.JsonElement;
+import com.google.gson.JsonParseException;
+import com.google.gson.JsonParser;
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.DataResult;
+import com.mojang.serialization.JsonOps;
+import com.mojang.serialization.Lifecycle;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import net.minecraft.registry.CombinedDynamicRegistries;
+import net.minecraft.registry.DynamicRegistryManager;
+import net.minecraft.registry.Registry;
+import net.minecraft.registry.RegistryKey;
+import net.minecraft.registry.RegistryWrapper;
+import net.minecraft.registry.tag.TagKey;
+import net.minecraft.resource.Resource;
+import net.minecraft.resource.ResourceFinder;
+import net.minecraft.resource.ResourceManager;
+import net.minecraft.resource.SinglePreparationResourceReloader;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.profiler.Profiler;
+
+import net.fabricmc.fabric.api.resource.IdentifiableResourceReloadListener;
+import net.fabricmc.fabric.api.tag.v1.TagAliasGroup;
+
+public final class TagAliasLoader extends SinglePreparationResourceReloader<Map<RegistryKey<? extends Registry<?>>, List<TagAliasLoader.Data>>> implements IdentifiableResourceReloadListener {
+	public static final Identifier ID = Identifier.of("fabric-tag-api-v1", "tag_alias_groups");
+
+	private static final Logger LOGGER = LoggerFactory.getLogger("fabric-tag-api-v1");
+	private final RegistryWrapper.WrapperLookup registries;
+
+	public TagAliasLoader(RegistryWrapper.WrapperLookup registries) {
+		this.registries = registries;
+	}
+
+	@Override
+	public Identifier getFabricId() {
+		return ID;
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	protected Map<RegistryKey<? extends Registry<?>>, List<TagAliasLoader.Data>> prepare(ResourceManager manager, Profiler profiler) {
+		Map<RegistryKey<? extends Registry<?>>, List<TagAliasLoader.Data>> dataByRegistry = new HashMap<>();
+		Iterator<RegistryKey<? extends Registry<?>>> registryIterator = registries.streamAllRegistryKeys().iterator();
+
+		while (registryIterator.hasNext()) {
+			RegistryKey<? extends Registry<?>> registryKey = registryIterator.next();
+			ResourceFinder resourceFinder = ResourceFinder.json(getDirectory(registryKey));
+
+			for (Map.Entry<Identifier, Resource> entry : resourceFinder.findResources(manager).entrySet()) {
+				Identifier resourcePath = entry.getKey();
+				Identifier groupId = resourceFinder.toResourceId(resourcePath);
+
+				try (Reader reader = entry.getValue().getReader()) {
+					JsonElement json = JsonParser.parseReader(reader);
+					Codec<TagAliasGroup<Object>> codec = TagAliasGroup.codec((RegistryKey<? extends Registry<Object>>) registryKey);
+
+					switch (codec.parse(JsonOps.INSTANCE, json)) {
```

Comment:
Shouldn't this be parsed with `RegistryOps`?

Reply:
The codec doesn't depend on `RegistryOps` since it doesn't parse registry entries, only tag *keys* (which don't depend on registry content).